### PR TITLE
add LAPACKFullMatrix::transpose()

### PR DESCRIPTION
**Pull request by davydden**
~~could someone with `lapack`on linux try the test? I ran on `macOS` and it fails.~~

TODO:

- [x] wait until Intel [clarifies documentation](https://software.intel.com/en-us/comment/1930653) of these functions. ~~So far I can't make it work for anything but square matrices.~~ EDIT: wrong docu on Inte's side.

p.s. looks like `openblas` has [cblas_?omatcopy](https://github.com/xianyi/OpenBLAS/blob/develop/cblas.h#L338-L345), so one could support this extension from other Blas providers as well.

### CODE REVIEW

**Comment by masterleinad** [comment]
The test passes for me.

**Comment by davydden** [comment]
> The test passes for me.

Thanks for checking! Presumably MKL is picked up during configure (you can check the new variable)?

**Comment by masterleinad** [comment]
No, `MKL` is not picked up.

**Comment by davydden** [comment]
> No, MKL is not picked up.

You would need to point to include directory with mkl headers (cmake’s Lapack). I do all manually (actually all is fetched correctly by Spack), set lapack as found and point to libs and include dirs.

**Comment by davydden** [comment]
both `openblas` and `intel-mkl` are working as expected. That's ready for review.

**Comment by drwells** [comment]
I would prefer that the new MKL wrapper functions go in a separate file: they are not, after all, LAPACK functions (and I would like to ultimately like to make that file 100% generated code).

**Comment by davydden** [comment]
Fair enough, how would you call that new header? MKL categorizes those functions as “BLAS-like extensions” https://software.intel.com/en-us/mkl-developer-reference-c-blas-like-extensions, openblas keeps its alternative in “cblas.h”.

Given that at least this specific function is implemented in MKL (mkl_omatcopy) and in openblas(cblas_domatcopy) under different names, perhaps we call the header “cblas_wrappers.h” and use cblas_omatcopy as the name of the wrapper?

**Comment by drwells** [comment]
It seems odd to name it after cblas when we only have MKL wrappers and checks at the moment. How about `blas_extension_wrappers.h` to be more generic?

**Comment by davydden** [comment]
Ok, will change.

**Comment by davydden** [comment]
@drwells I updated the PR

File: include/deal.II/lac/blas_extension_templates.h [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+#  include <cfenv>
+#endif
+
+// Intel-MKL specific functions
+#ifdef DEAL_II_LAPACK_WITH_MKL
+// see
+// https://software.intel.com/en-us/mkl-windows-developer-guide-using-complex-types-in-c-c
+#  define MKL_Complex8 std::complex<float>
+#  define MKL_Complex16 std::complex<double>
+#  include "mkl_trans.h"
```
Comment (drwells) [inline include/deal.II/lac/blas_extension_templates.h]: Should this be `include <mkl_trans.h>`?

Comment (davydden) [reply include/deal.II/lac/blas_extension_templates.h]: fixed

**Comment by davydden** [comment]
@dealii/dealii can anyone review this Lapack PR? I would like to have this merge and go on with other developments...

**Pull request merged**